apps/settings/js/modules/feature_model.js

```diff
@@ -0,0 +1,99 @@
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+  var Observable = require('modules/mvvm/observable');
+
+  /*
+   * A model needs to be created with an options object with the following
+   * three functions:
+   *
+   * onInit is called once with a function as the first argument when the
+   * model is initializing. The function is used for reporting the state
+   * changes. The function should return a initial state immediately.
+   *
+   * onSetState is called with a desired state. Do the operation required for
+   * switching to the desired state and return a promise wrapping the state
+   * after the operation no matter it resolves or rejects.
+   *
+   * onGetState is called for requesting the current state.
+   *
+   * @class FeatureModel
+   * @param {Object} options
+   * @param {Function} options.onInit
+   * @param {Function} options.onSetState
+   * @param {Function} options.onGetState
+   * @returns {FeatureModel}
+   */
+  var FeatureModel = Module.create(function FeatureModel(options) {
+    this.super(Observable).call(this);
+
+    this._init(options);
+  }).extend(Observable);
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'targetState', {
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'currentState', {
+    readonly: true,
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'transitioning', {
+    readonly: true,
+    value: false
+  });
+
+  FeatureModel.prototype._init = function(options) {
+    // XXX: do more checks here
+    this._options = options;
+
+    this.targetState = this._options.onInit((realState) => {
```

Comment:
`onInit` is the best name from the aspect of options because it is only called once and is used for initializing the options object (just like what we did in `SettingsPanel`). I guess we can improve this by adding more documentations.